File: Leetcode/Stone Game/Stone Game.py
"""
A python program to implement Stone Game
Alex and Lee play a game with piles of stones.  There are an even number of piles arranged in a row, and each pile has a positive integer number of stones piles[i].
The objective of the game is to end with the most stones.  The total number of stones is odd, so there are no ties.
Alex and Lee take turns, with Alex starting first.  Each turn, a player takes the entire pile of stones from either the beginning or the end of the row.  This continues until there are no more piles left, at which point the person with the most stones wins.
Assuming Alex and Lee play optimally, return True if and only if Alex wins the game.

"""

class Solution(object):
    def stoneGame(self, piles):
        """
        :type piles: List[int]
        :rtype: bool
        """
        if len(piles) % 2 ==0 or len(piles) == 1:
            return True
        
        dp = [0] * len(piles);
        for i in reversed(xrange(len(piles))):
            dp[i] = piles[i]
            for j in xrange(i+1, len(piles)):
                dp[j] = max(piles[i] - dp[j], piles[j] - dp[j-1])
        return dp[-1] >= 0 
      
"""
Sample Input/Output
Example 1:
Input: piles = [5,3,4,5]
Output: true

"""

File: Leetcode/Stone Game/Stone_Game.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution {
   public:
   bool stoneGame(vector<int>& piles)
   {
       //calculating size of the array piles
      int n = piles.size();

      //declaring a matrix dp if size n*n 
      vector < vector <int> > dp(n,vector <int>(n));

      //declaring an array of size n+1
      vector <int> pre(n + 1);

      //taking stones 
      for(int i = 0; i < n; i++)
      {
         pre[i + 1] = pre[i] + piles[i];
      }
      for(int l = 2; l <= n; l++)
      {
         for(int i = 0, j = l - 1; j < n; i++, j++)
         {
            dp[i][j] = max(piles[j]+pre[j]-pre[i]-dp[i][j-1],piles[i]+pre[i+2]- pre[j]+dp[i+1][j]);
         }
      }

      //returning true if only Alex wins
      return dp[0][n - 1] > dp[0][n - 1] - pre[n];
   }
};
